### DISCUSSION TITLE
What's the best way of doing dynamic filtering?

### DISCUSSION THREAD

**Discussion by kitsunde** [uwdata/arquero#298] [Q&A]
I need to filter a table based on a dynamic filter.

```js
const filters = [
  {column: 'a', min: 0, max: 10},
  {column: 'b', min: 2, max: 7}
]
```

So since I can't reference it from the outside I pass it into params:

```js
table.params({ filters })
```

However various ways of looping isn't possible:

```js
table.filter((d, $) => {
  for (const filter of $.filters) {
    // ...
  }
})
```

> For-of loops not allowed

```js
table.filter((d, $) => {
  for (let i = 0; i < $.filters.length; i++) {
    // ...
  }
})
```

> For loops not allowed

And of course, using an array method is also not:

```js
table.filter((d, $) => {
  $.filters.every(() => { /*....*/ });
})
```

>  Table expressions do not support closures.

So then I can pass params and call filter individually:

```js
filters.forEach((filter) => {
  table = table.params({filter}).filter((d, $) => { /**/ })
})
```

Now this works, but I am a bit concerned how this would actually be handled internally, would each filter need to pass over the data one at a time? Would overriding the same attribute repeatedly and then filtering on those properties be wise?

The alternative seem to be:

* Build the filters as a string (seems like the correct option?)
* Escape the function to be able to use closures / loops.
* Build a second table with conditions and do a join against that
* Dynamically add a new function?

**Comment by kitsunde** [comment]
Additionally, if I build the filters as a string. Is it significantly better to apply all conditions in one large string with many conditions, or does it not matter to call filter multiple times?

**Reply by stuvjordan-uroc** [reply]
Hi @kitsunde .  I don't have an answer, but I have the same question.  Just want to say I feel your pain.  Wish the maintainers would follow up on this.
